File: packages/cli/src/commands/simulate.ts
/**
 * `simforge simulate <instance> [--trace out.trace.json.gz]`.
 *
 * One native engine pass. Feasibility guards are collected, not thrown: a
 * scenario that fails a guard is still worth simulating — the resulting
 * metrics are how you tell "the runway is 8 m short" from "200 m short".
 */

import path from 'node:path';

import type { SimTrace } from '@simforge-oss/engine';
import { parseTrace, runSimulation } from '@simforge-oss/engine/node';
import { loadMap, readInstance, writeTraceFile } from '@simforge-oss/compiler/node';

import { EXIT } from '../errors.js';
import { emit, emitLines, fixed, pad } from '../output.js';

export interface SimulateOptions {
  readonly file: string;
  readonly trace?: string | undefined;
  readonly pretty: boolean;
}

/** The metrics block every command that runs the engine prints. */
export function metricsSummary(trace: SimTrace): Record<string, unknown> {
  const m = trace.metrics;
  return {
    minTTC: m.minTTC
      ? { value: round(m.minTTC.value), t: round(m.minTTC.t), pair: m.minTTC.pair }
      : null,
    minDistance: m.minDistance.map((d) => ({
      pair: d.pair,
      minDistanceM: round(d.minDistanceM),
      t: round(d.t),
    })),
    requiredDecelMax: Object.fromEntries(
      Object.entries(m.requiredDecelMax).map(([k, v]) => [k, round(v)]),
    ),
    revealToConflict: m.revealToConflict
      ? {
          observer: m.revealToConflict.observer,
          target: m.revealToConflict.target,
          value: round(m.revealToConflict.value),
          firstBlockedT: round(m.revealToConflict.firstBlockedT),
          losOpenT: round(m.revealToConflict.losOpenT),
          conflictT: round(m.revealToConflict.conflictT),
          pair: m.revealToConflict.pair,
          occluderId: m.revealToConflict.occluderId ?? null,
          relevantOccluderIds: m.revealToConflict.relevantOccluderIds,
        }
      : null,
    declaredOcclusion: (m.declaredOcclusion ?? []).map((entry) => ({
      observer: entry.observer,
      target: entry.target,
      pair: entry.pair,
      occluderId: entry.occluderId ?? null,
      relevantOccluderIds: entry.relevantOccluderIds,
      status: entry.status,
      firstBlockedT: entry.firstBlockedT === null ? null : round(entry.firstBlockedT),
      losOpenT: entry.losOpenT === null ? null : round(entry.losOpenT),
      conflictT: entry.conflictT === null ? null : round(entry.conflictT),
      revealToConflictS: entry.revealToConflictS === null ? null : round(entry.revealToConflictS),
    })),
    occluderIneffective: (m.occluderIneffective ?? []).map((entry) => ({
      observer: entry.observer,
      target: entry.target,
      pair: entry.pair,
      conflictT: round(entry.conflictT),
      firstBlockedT: entry.firstBlockedT === undefined ? null : round(entry.firstBlockedT),
      occluderId: entry.occluderId ?? null,
      relevantOccluderIds: entry.relevantOccluderIds,
      reason: entry.reason,
    })),
    collisions: m.collisions,
    triggerNeverFired: m.triggerNeverFired,
    clippedCriticality: m.clippedCriticality,
    ticksSimulated: m.ticksSimulated,
  };
}

function round(value: number): number {
  return Math.round(value * 1000) / 1000;
}

export async function simulate(options: SimulateOptions): Promise<number> {
  const instance = await readInstance(options.file);
  const bundle = await loadMap(instance.input.mapId);
  const result = runSimulation(instance.input, { graph: bundle.graph });
  const trace = parseTrace(result.trace);
  const digest = trace.digest();

  if (options.trace) {
    await writeTraceFile(options.trace, trace);
  }

  const payload = {
    file: options.file,
    mapId: instance.input.mapId,
    header: result.trace.header,
    traceDigest: digest,
    metrics: metricsSummary(result.trace),
    events: countEvents(result.trace),
    issues: result.issues,
    arrival: result.arrival,
    trace: options.trace ? path.resolve(options.trace) : null,
  };

  if (!options.pretty) {
    emit(payload, options);
  } else {
    const m = result.trace.metrics;
    const lines = [
      `${options.file} on ${instance.input.mapId} — ${result.trace.ticks.t.length} recorded ticks, engine ${result.trace.header.engineVersion}`,
      `inputHash ${result.trace.header.inputHash.slice(0, 16)}…  traceDigest ${digest.slice(0, 16)}…`,
      '',
      `minTTC              ${m.minTTC ? `${fixed(m.minTTC.value)} s at t=${fixed(m.minTTC.t)} s (${m.minTTC.pair.join(' / ')})` : '— (no pair ever closed)'}`,
      `minDistance         ${m.minDistance
        .map((d) => `${d.pair.join('/')}: ${fixed(d.minDistanceM)} m @ ${fixed(d.t)} s`)
        .join(', ') || '—'}`,
      `requiredDecelMax    ${Object.entries(m.requiredDecelMax)
        .map(([k, v]) => `${k}: ${fixed(v)}`)
        .join(', ')}`,
      `revealToConflict    ${m.revealToConflict ? `${fixed(m.revealToConflict.value)} s (${m.revealToConflict.occluderId ?? 'any'}: blocked ${fixed(m.revealToConflict.firstBlockedT)} s, LOS opened ${fixed(m.revealToConflict.losOpenT)} s, conflict ${fixed(m.revealToConflict.conflictT)} s; members ${m.revealToConflict.relevantOccluderIds.join(',') || '—'})` : '—'}`,
      `occluderIneffective ${(m.occluderIneffective ?? []).map((e) => `${e.pair.join(' / ')} via ${e.occluderId ?? 'any'} (${e.reason}; members ${e.relevantOccluderIds.join(',') || '—'})`).join(', ') || '—'}`,
      `collisions          ${m.collisions.length}`,
      `triggerNeverFired   ${m.triggerNeverFired.join(', ') || '—'}`,
      `clippedCriticality  ${m.clippedCriticality}`,
    ];
    if (result.issues.length > 0) {
      lines.push('', 'issues:');
      for (const i of result.issues) lines.push(`  ${pad(i.severity, 9)}${pad(i.code, 26)}${i.reason}`);
    }
    if (options.trace) lines.push('', `trace: ${path.resolve(options.trace)}`);
    emitLines(lines);
  }

  return result.issues.some((i) => i.severity === 'error') ? EXIT.validationFindings : EXIT.ok;
}

function countEvents(trace: SimTrace): Record<string, number> {
  const out: Record<string, number> = {};
  for (const event of trace.events) out[event.kind] = (out[event.kind] ?? 0) + 1;
  return out;
}

File: native/crates/simforge-authoring/src/simulate.rs
//! `simulate <instance> [--trace <file>]`: one engine pass over an instance
//! file (or a bare SimScenarioInput). Feasibility guards are collected, not
//! thrown: a scenario that fails a guard is still worth simulating, because
//! the metrics are how "the runway is 8 m short" differs from "200 m short".

use std::path::Path;

use simforge_compiler::CompileError;

use crate::engine::{quantized_json, reparse_trace, simulate, write_trace_gz};
use crate::jsvalue::JsValue;
use crate::maps::MapRoot;
use crate::metrics::metrics_summary;
use crate::readers::read_instance;

/// The `simulate` document and whether the run had no error-severity issue.
pub fn run_simulate_instance(
    root: &MapRoot,
    file: &Path,
    file_arg: &str,
    trace_out: Option<&Path>,
) -> Result<(JsValue, bool), CompileError> {
    let instance = read_instance(file)?;
    let map = root.load(&instance.input.map_id)?;
    let result = simulate(&instance.input, &map)?;
    let handle = reparse_trace(&result.trace)?;
    let digest = handle
        .digest()
        .map_err(|e| CompileError::internal(format!("trace digest: {e}")))?;
    let written = match trace_out {
        Some(out) => {
            write_trace_gz(out, &quantized_json(&handle)?)?;
            JsValue::String(crate::paths::resolve(out).display().to_string())
        }
        None => JsValue::Null,
    };
    // `countEvents`: occurrences per event kind, in first-seen order.
    let mut counts: Vec<(String, JsValue)> = Vec::new();
    for event in JsValue::from_serialize(&result.trace.events)?
        .as_array()
        .unwrap_or(&[])
    {
        let kind = event
            .get("kind")
            .and_then(JsValue::as_str)
            .unwrap_or("undefined")
            .to_owned();
        match counts.iter_mut().find(|(k, _)| *k == kind) {
            Some((_, JsValue::Number(n))) => *n += 1.0,
            _ => counts.push((kind, JsValue::Number(1.0))),
        }
    }
    let errors = result
        .issues
        .iter()
        .any(|i| i.severity == simforge_core::error::SimIssueSeverity::Error);
    let doc = JsValue::object(vec![
        ("file".into(), JsValue::String(file_arg.to_owned())),
        (
            "mapId".into(),
            JsValue::String(instance.input.map_id.clone()),
        ),
        (
            "header".into(),
            JsValue::from_serialize(&result.trace.header)?,
        ),
        ("traceDigest".into(), JsValue::String(digest)),
        (
            "metrics".into(),
            JsValue::from_serialize(&metrics_summary(&result.trace.metrics))?,
        ),
        ("events".into(), JsValue::object(counts)),
        ("issues".into(), JsValue::from_serialize(&result.issues)?),
        ("arrival".into(), JsValue::from_serialize(&result.arrival)?),
        ("trace".into(), written),
    ]);
    Ok((doc, !errors))
}
